org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend

```diff
@@ -3587,6 +3710,108 @@ class RenameParams extends TextDocumentPositionAndWorkDoneProgressParams {
 	}
 }
 
+@Beta
+@JsonRpcData
+class SemanticTokensLegend {
+
+	/**
+	 * The token types that the client supports.
+	 */
+	@NonNull
+	List<String> tokenTypes;
+
+	/**
+	 * The token modifiers that the client supports.
+	 */
+	@NonNull
+	List<String> tokenModifiers;
+
+	new(@NonNull List<String> tokenTypes, @NonNull List<String> tokenModifiers) {
+		this.tokenTypes = Preconditions.checkNotNull(tokenTypes, 'tokenTypes')
+		this.tokenModifiers = Preconditions.checkNotNull(tokenModifiers, 'tokenModifiers')
+	}
+
+}
+
+@Beta
+@JsonRpcData
+class SemanticTokensServerFull {
+
+	/**
+	* The server supports deltas for full documents.
+	*/
+	Boolean delta
+
+	new() {
+	}
+
+	new(Boolean delta) {
+		this.delta = delta
+	}
+}
+
+@Beta
+@JsonRpcData
+class SemanticTokensWithRegistrationOptions extends AbstractWorkDoneProgressOptions {
+
+	/**
+	 * The legend used by the server
+	 */
+	@NonNull
+	SemanticTokensLegend legend
+
+	/**
+	 * Server supports providing semantic tokens for a specific range
+	 * of a document.
+	 */
+	Either<Boolean, Object> range
+
+	/**
+	 * Server supports providing semantic tokens for a full document.
+	 */
+	Either<Boolean, SemanticTokensServerFull> full
+
+	/**
+	 * A document selector to identify the scope of the registration. If set to null
+	 * the document selector provided on the client side will be used.
+	 */
+	List<DocumentFilter> documentSelector
+
+	new(@NonNull SemanticTokensLegend legend) {
+		this.legend = Preconditions.checkNotNull(legend, 'legend')
+	}
+
+	new(@NonNull SemanticTokensLegend legend, Boolean full) {
+		this(legend)
+		this.full = full
+	}
+
+	new(@NonNull SemanticTokensLegend legend, SemanticTokensServerFull full) {
+		this(legend)
+		this.full = full
+	}
+
+	new(@NonNull SemanticTokensLegend legend, Boolean full, Boolean range) {
+		this(legend)
+		this.full = full
+		this.range = range
+	}
+
+	new(@NonNull SemanticTokensLegend legend, SemanticTokensServerFull full, Boolean range) {
+		this(legend)
+		this.full = full
+		this.range = range
+	}
+
+	new(@NonNull SemanticTokensLegend legend, SemanticTokensServerFull full, Boolean range, List<DocumentFilter> documentSelector) {
```

Comment:
The specification says it's either a `SemanticTokensOptions` (with `legend`)  or a `SemanticTokensRegistrationOptions` (with `documentSelector`), but not both.

Please also add a no-arguments constructor. In some contexts it's easier to construct the protocol classes with setters, e.g. with Xtend:
```xtend
options = new SemanticTokensWithRegistrationOptions => [
    legend = ...
    range = ...
]

Reply:
I don't get it, what do you suggest?
I thought that [this comment](https://github.com/eclipse/lsp4j/pull/446#discussion_r493545332) suggested merging both classes

Reply:
`SemanticTokensRegistrationOptions` extends `SemanticTokensOptions`, so `legend` would be in both.

Reply:
`SemanticTokensRegistrationOptions` extends `TextDocumentRegistrationOptions`, which has a `documentSelector`, but no `legend`.

https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#textDocument_semanticTokens

> I thought that this comment suggested merging both classes

Yes, but merging the classes is just a workaround because Java's type system is not as flexible as TypeScript. We use a single class to parse / serialize both alternatives of `SemanticTokensOptions | SemanticTokensRegistrationOptions`.

Reply:
See above, I didn't realize that SemanticTokensRegistrationOptions is actually meant to extend SemanticTokensOptions.  Then the current constructors can be kept, but please add one without arguments:
```xtend
new() {
}
```

Reply:
Added. I didn't added a empty constructor before because both `SemanticTokensOptions` and `SemanticTokensRegistrationOptions` have non-null `legend` filed, so I thought that it didn't make sense to add a constructor that doesn't initialize the `legend` field.